File: Incubation/trunk/psim/src/org/space/oss/psim/time/TimeServiceBase.java
package org.space.oss.psim.time;

import java.util.ArrayList;
import java.util.List;

import org.space.oss.psim.Config;
import org.space.oss.psim.PSim;
import org.space.oss.psim.TimeService;
import org.space.oss.psim.TimeServiceObserver;

public class TimeServiceBase implements TimeService 
{
	protected PSim psim_;
	protected long currentTime_;
	protected List<TimeServiceObserver> observers_;
	
	public TimeServiceBase()
	{
	}

	@Override
	public void init(PSim psim, Config cfg) 
	{
		psim_ = psim;
		currentTime_ = 0;
		observers_ = new ArrayList<TimeServiceObserver>();		
	}

	@Override
	public void shutdown() 
	{
	}

	
	@Override
	public long getCurrentTime() { return currentTime_; }

	@Override
	public void setCurrentTime(long t) 
	{ 
		currentTime_ = t; 
		notifyCurrentTimeChanged();
	}

	protected void notifyCurrentTimeChanged()
	{
		for (TimeServiceObserver o : observers_)
			o.handleCurrentTime(currentTime_);
	}
	
	@Override
	public void addObserver(TimeServiceObserver o) 
	{
		observers_.add(o);
	}

	@Override
	public void removeObserver(TimeServiceObserver o) 
	{
		observers_.remove(o);
	}
}

File: Incubation/trunk/psim-scala/src/main/scala/org/space/oss/psim/time/TimeServiceBase.scala
package org.space.oss.psim.time

import org.space.oss.psim.TimeService
import org.space.oss.psim.PSimUtil

class TimeServiceBase extends TimeService {
  var currentTime=0L
  var timeListeners = scala.collection.mutable.ListBuffer[(Long)=>Unit]()
  
  override def getCurrentTime = currentTime
  
  override def setCurrentTime(t:Long) {
    println("Advancing to time "+PSimUtil.formatTime(t))
    currentTime = t
    for (listenTo <- timeListeners)
    	listenTo(currentTime)    
  }
  
  override def addObserver(o:(Long)=>Unit) { timeListeners += o }
}